src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,35 +1,86 @@
 package com.es.phoneshop.web;
 
 
+import com.es.phoneshop.model.card.Cart;
+import com.es.phoneshop.model.card.CartService;
+import com.es.phoneshop.model.card.HttpSessionCartService;
+import com.es.phoneshop.model.exceptions.OutOfStockException;
 import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
 import com.es.phoneshop.model.product.ProductDao;
+import com.es.phoneshop.model.product.ProductHistory;
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.Locale;
 
 
 public class ProductDetailsPageServlet extends HttpServlet {
 
-    private ProductDao products;
+    private ProductDao productDao;
+    private CartService cartService;
+    private ProductHistory productHistory;
 
     @Override
     public void init() {
-        products = ArrayListProductDao.getInstance();
+        productDao = ArrayListProductDao.getInstance();
+        cartService = HttpSessionCartService.getInstance();
+        productHistory = ProductHistory.getInstance();
     }
 
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String code = request.getPathInfo().substring(1);
+        String code = getCode(request);
         try {
-            request.setAttribute("product", products.getProduct(code));
-            request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
+            Product product = productDao.getProduct(code);
+            showPage(request, response, product);
+            if (productHistory.getRecentProducts().size() > 2 & productHistory.save(product)) {
```

Comment:
Use && in boolean operations

Reply:
\+ Where is productHistory supposed to take recently viewed products from? They must be specific for each customer

Reply:
\+ Maybe it makes sense to move all this logic (checking the history size, deleting from history, etc) to productHistoryService itself